pallene/checker.lua

```diff
@@ -636,29 +638,47 @@ function FunChecker:check_var(var)
         elseif cname._tag == "checker.Name.Function" then
             var._type = self.p.module.functions[cname.id].typ
         elseif cname._tag == "checker.Name.Builtin" then
-            var._type = builtins[cname.name].typ
+            var._type = builtins.functions[cname.name].typ
         elseif cname._tag == "checker.Name.Module" then
             -- The type for a module is a table.
-            var._type = builtin_modules[cname.name]
+           var._type = builtins.modules[cname.name]
         else
-            error("Unexpected condition: cname._tag == " .. cname._tag)
+            error("impossible")
         end
 
+        -- print(inspect(var))
+
     elseif tag == "ast.Var.Dot" then
-        var.exp = self:check_exp_synthesize(var.exp)
-        local ind_type = var.exp._type
-        if not types.is_indexable(ind_type) then
-            type_error(var.loc,
-                "trying to access a member of value of type '%s'",
-                types.tostring(ind_type))
-        end
-        local field_type = types.indices(ind_type)[var.name]
-        if not field_type then
-            type_error(var.loc,
-                "field '%s' not found in type '%s'",
-                var.name, types.tostring(ind_type))
+        -- print(inspect(var))
+        -- print('---------------------')
+        if builtins.modules[var.exp.var.name] ~= nil then
```

Comment:
Before you access `exp.var` you need to check that the `exp` in question is an `ast.Exp.Var`. This line of code will blow up if the expression on the left of the dot is not a variable name. For example `xs[i].name`

By the way, if the existing test suite is not catching this bug then please add a test case that ensures that it does :)

Reply:
Another thing: you can leave out the `~= nil`. In the rest of the code base we just use `if set[key] then` instead of `if set[key] ~= nil then`

Reply:
Done.